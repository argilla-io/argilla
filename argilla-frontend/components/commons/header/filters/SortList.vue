<!--
  - coding=utf-8
  - Copyright 2021-present, the Recognai S.L. team.
  -
  - Licensed under the Apache License, Version 2.0 (the "License");
  - you may not use this file except in compliance with the License.
  - You may obtain a copy of the License at
  -
  -     http://www.apache.org/licenses/LICENSE-2.0
  -
  - Unless required by applicable law or agreed to in writing, software
  - distributed under the License is distributed on an "AS IS" BASIS,
  - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  - See the License for the specific language governing permissions and
  - limitations under the License.
  -->

<template>
  <div>
    <div class="filters--scrollable">
      <SortFilter
        v-for="index in numberOfSortFields"
        :key="index"
        class="sort"
        :selected-field="selectedFields[index - 1]"
        :sort-options="filteredSortOptions"
        @removeSortField="onRemoveSortField(index)"
        @addSortField="onAddSortField(index, ...arguments)"
      />
    </div>
    <base-button
      v-if="selectedFields.length === numberOfSortFields"
      class="sort__add-button secondary light small"
      @click="addNewField"
      >+ Add another field</base-button
    >
    <div class="sort__buttons" v-if="sort.length || selectedFields.length">
      <base-button class="primary outline" @click="cancel">Cancel</base-button>
      <base-button class="primary" @click="apply">Sort</base-button>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    sortOptions: {
      type: Array,
      required: true,
    },
    sort: {
      type: Array,
    },
  },
  data: () => {
    return {
      numberOfSortFields: 1,
      selectedFields: [],
    };
  },
  computed: {
    filteredSortOptions() {
      const formatOptions = this.sortOptions.map((opt) => {
        if (opt.group.toLowerCase() === "metadata") {
          return {
            ...opt,
            id: `metadata.${opt.name}`,
            name: `Metadata.${opt.name}`,
          };
        }
        return opt;
      });
      return formatOptions.filter(
        (opt) =>
          !this.selectedFields.some((field) => opt.id.toString() === field.id)
      );
    },
  },
  mounted() {
    this.numberOfSortFields = this.sort.length === 0 ? 1 : this.sort.length;
    this.selectedFields = [...this.sort];
  },
  methods: {
    addNewField() {
      this.numberOfSortFields += 1;
    },
    onRemoveSortField(index) {
      if (this.selectedFields.length === 1) {
        this.$emit("sortBy", []);
      } else {
        this.selectedFields.splice(index - 1, 1);
      }
      if (this.numberOfSortFields > 1) {
        this.numberOfSortFields -= 1;
      }
    },
    onAddSortField(index, option, direction) {
      const item = {
        ...option,
        id:
          option.group.toLowerCase() === "metadata"
            ? "metadata." + option.key
            : option.key,
        name: option.name,
        order: direction,
      };
      this.selectedFields.splice(index - 1, 1, item);
    },
    apply() {
      this.$emit("sortBy", this.selectedFields);
    },
    cancel() {
      this.$emit("closeSort");
    },
  },
};
</script>

<style lang="scss" scoped>
.sort {
  margin-bottom: 1em;
  &__buttons {
    display: flex;
    margin-top: 1.5em;
    margin-bottom: 10px;
    & > * {
      display: block;
      width: 100%;
      &:last-child {
        margin-left: $base-space;
      }
    }
  }
  &__add-button {
    margin-top: 1em;
  }
}
</style>
