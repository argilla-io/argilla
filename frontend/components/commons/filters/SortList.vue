<template>
  <div>
    <SortFilter
      v-for="index in numberOfSortFields"
      :key="index"
      class="sort"
      :selected-field="selectedFields[index - 1]"
      :sort-options="filteredSortOptions"
      @removeSortField="onRemoveSortField(index)"
      @addSortField="onAddSortField(index, ...arguments)"
    />
    <a
      v-if="selectedFields.length === numberOfSortFields"
      class="sort__add-button"
      href="#"
      @click="addNewField"
      >add field +</a
    >
    <re-button
      v-if="selectedFields.length"
      class="discarded sort__button button-primary--small"
      @click="apply"
      >Apply</re-button
    >
  </div>
</template>

<script>
export default {
  props: {
    sortOptions: {
      type: Array,
      required: true,
    },
  },
  data: () => {
    return {
      numberOfSortFields: 1,
      selectedFields: [],
    };
  },
  computed: {
    filteredSortOptions() {
      return this.sortOptions.filter(
        (opt) =>
          !this.selectedFields.some((field) => opt.id.toString() === field.id)
      );
    },
  },
  methods: {
    addNewField() {
      this.numberOfSortFields += 1;
    },
    onRemoveSortField(index) {
      this.selectedFields.splice(index - 1, 1);
      if (this.numberOfSortFields > 1) {
        this.numberOfSortFields -= 1;
      }
      this.apply();
    },
    onAddSortField(index, option, direction) {
      const item = {
        id: option.group == "Metadata" ? "metadata." + option.id : option.id,
        name: option.name,
        order: direction,
      };
      this.selectedFields.splice(index - 1, 1, item);
    },
    apply() {
      this.$emit("sortBy", this.selectedFields);
    },
  },
};
</script>

<style lang="scss" scoped>
.sort {
  margin-bottom: 0.5em;
  &__button {
    margin-left: auto;
    margin-right: 0;
    margin-bottom: 0;
  }
  &__add-button {
    margin-top: 1em;
    color: $primary-color;
    outline: none;
    text-decoration: none;
    display: block;
  }
}
</style>
