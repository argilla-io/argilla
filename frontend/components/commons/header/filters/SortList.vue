<!--
  - coding=utf-8
  - Copyright 2021-present, the Recognai S.L. team.
  -
  - Licensed under the Apache License, Version 2.0 (the "License");
  - you may not use this file except in compliance with the License.
  - You may obtain a copy of the License at
  -
  -     http://www.apache.org/licenses/LICENSE-2.0
  -
  - Unless required by applicable law or agreed to in writing, software
  - distributed under the License is distributed on an "AS IS" BASIS,
  - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  - See the License for the specific language governing permissions and
  - limitations under the License.
  -->

<template>
  <div>
    <SortFilter
      v-for="index in numberOfSortFields"
      :key="index"
      class="sort"
      :selected-field="selectedFields[index - 1]"
      :sort-options="filteredSortOptions"
      @removeSortField="onRemoveSortField(index)"
      @addSortField="onAddSortField(index, ...arguments)"
    />
    <a
      v-if="selectedFields.length === numberOfSortFields"
      class="sort__add-button"
      href="#"
      @click="addNewField"
      >+ Add another field</a
    >
    <div class="sort__buttons">
      <re-button
        class="button-tertiary--small button-tertiary--outline"
        @click="cancel"
        >Cancel</re-button
      >
      <re-button class="button-primary--small" @click="apply">Apply</re-button>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    sortOptions: {
      type: Array,
      required: true,
    },
    sort: {
      type: Array,
    },
  },
  data: () => {
    return {
      numberOfSortFields: 1,
      selectedFields: [],
    };
  },
  computed: {
    filteredSortOptions() {
      return this.sortOptions.filter(
        (opt) =>
          !this.selectedFields.some((field) => opt.id.toString() === field.id)
      );
    },
  },
  mounted() {
    this.numberOfSortFields = this.sort.length === 0 ? 1 : this.sort.length;
    this.selectedFields = [...this.sort];
  },
  methods: {
    addNewField() {
      this.numberOfSortFields += 1;
    },
    onRemoveSortField(index) {
      this.selectedFields.splice(index - 1, 1);
      if (this.numberOfSortFields > 1) {
        this.numberOfSortFields -= 1;
      }
    },
    onAddSortField(index, option, direction) {
      const item = {
        ...option,
        id:
          option.group.toLowerCase() === "metadata"
            ? "metadata." + option.key
            : option.key,
        name: option.name,
        order: direction,
      };
      this.selectedFields.splice(index - 1, 1, item);
    },
    apply() {
      this.$emit("sortBy", this.selectedFields);
    },
    cancel() {
      this.$emit("closeSort");
    },
  },
};
</script>

<style lang="scss" scoped>
.sort {
  margin-bottom: 0.5em;
  &__buttons {
    display: flex;
    margin-top: 1.5em;
    button {
      margin-bottom: 0;
    }
    button:last-child {
      margin-right: auto;
      margin-left: 0.5em;
    }
  }
  &__add-button {
    margin-top: 1em;
    color: $primary-color;
    outline: none;
    text-decoration: none;
    display: block;
  }
}
</style>
