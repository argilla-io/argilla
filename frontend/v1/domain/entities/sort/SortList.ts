import { Metadata } from "../metadata/Metadata";
import { Question } from "../question/Question";

const SORT_ASC = "asc";
const SORT_DESC = "desc";

type SortOrderOptions = "asc" | "desc";

type SortOptions = "metadata" | "record" | "suggestion" | "response";

type SortProperty = "score" | "value";

type SortId = string;

export interface SortSearch {
  entity: SortOptions;
  name: string;
  property?: SortProperty;
  order: SortOrderOptions;
}

abstract class Sort {
  public sort: SortOrderOptions = SORT_ASC;

  constructor(
    public readonly key: SortOptions,
    public readonly property?: SortProperty
  ) {}

  toggleSort() {
    this.sort = this.sort === SORT_ASC ? SORT_DESC : SORT_ASC;
  }

  abstract get name(): string;

  abstract get title(): string;

  get id(): SortId {
    return this.property
      ? `${this.key}.${this.name}.${this.property}`
      : `${this.key}.${this.name}`;
  }

  get group(): string {
    return this.property ? `${this.key}.${this.property}` : this.key;
  }

  get canSort(): boolean {
    return true;
  }

  get tooltip() {
    return this.title;
  }
}
type SortIdentifier = Pick<Sort, "key" | "name" | "property">;

class MetadataSort extends Sort {
  constructor(private readonly metadata: Metadata) {
    super("metadata");
  }

  get name() {
    return this.metadata.name;
  }

  get title() {
    return this.metadata.title;
  }

  get canSort(): boolean {
    return this.metadata.hasValues;
  }
}

class SuggestionScoreSort extends Sort {
  constructor(private readonly question: Question) {
    super("suggestion", "score");
  }

  get name() {
    return this.question.name;
  }

  get title() {
    return this.question.name;
  }

  get tooltip() {
    return this.question.title;
  }
}

class SuggestionValueSort extends Sort {
  constructor(private readonly question: Question) {
    super("suggestion", "value");
  }

  get name() {
    return this.question.name;
  }

  get title() {
    return this.question.name;
  }

  get tooltip() {
    return this.question.title;
  }
}

class ResponseValueSort extends Sort {
  constructor(private readonly question: Question) {
    super("response");
  }

  get name() {
    return this.question.name;
  }

  get title() {
    return this.question.name;
  }

  get tooltip() {
    return this.question.title;
  }
}

class RecordSort extends Sort {
  constructor(public readonly name: string, public readonly title = name) {
    super("record");
  }
}

export class SortList {
  private categoriesSorts: Sort[] = [];
  private selectedCategories: Sort[] = [];
  private latestCommit: SortSearch[] = [];

  constructor(metadata: Metadata[], questions: Question[]) {
    this.categoriesSorts.push(new RecordSort("inserted_at"));
    this.categoriesSorts.push(new RecordSort("updated_at"));

    questions?.forEach((q) =>
      this.categoriesSorts.push(new SuggestionScoreSort(q))
    );

    questions
      ?.filter((q) => q.isRatingType)
      .forEach((q) => this.categoriesSorts.push(new SuggestionValueSort(q)));

    questions
      ?.filter((q) => q.isRatingType)
      .forEach((q) => this.categoriesSorts.push(new ResponseValueSort(q)));

    metadata?.forEach((metadata) =>
      this.categoriesSorts.push(new MetadataSort(metadata))
    );
  }

  get selected() {
    return this.selectedCategories;
  }

  get noSelected() {
    return this.categoriesSorts.filter(
      (metadata) => !this.selectedCategories.includes(metadata)
    );
  }

  select(sort: SortIdentifier) {
    const found = this.findByCategory(sort);

    if (found) {
      this.selectedCategories.push(found);
    }
  }

  unselect(sort: SortIdentifier) {
    const indexOf = this.findIndexSelectedCategory(sort);

    if (indexOf > -1) {
      this.selectedCategories.splice(indexOf, 1);
    }
  }

  replace(actualSort: SortIdentifier, newSort: SortIdentifier) {
    const newCategoryFound = this.findByCategory(newSort);

    const indexOf = this.findIndexSelectedCategory(actualSort);

    if (indexOf > -1) {
      this.selectedCategories.splice(indexOf, 1, newCategoryFound);
    }
  }

  clear() {
    this.selectedCategories = [];
  }

  toggleSort(sort: SortIdentifier) {
    const found = this.findByCategory(sort);

    if (found) found.toggleSort();
  }

  get hasChanges() {
    return this.hasDifferencesWith(this.createSortCriteria());
  }

  hasDifferencesWith(compare: SortSearch[]) {
    return JSON.stringify(this.latestCommit) !== JSON.stringify(compare);
  }

  commit(): SortSearch[] {
    this.latestCommit = this.createSortCriteria();

    return this.latestCommit;
  }

  private createSortCriteria(): SortSearch[] {
    return this.selectedCategories.map(
      ({ key: entity, property, name, sort: order }) => {
        if (property)
          return {
            entity,
            name,
            property,
            order,
          };

        return {
          entity,
          name,
          order,
        };
      }
    );
  }

  complete(sort: SortSearch[]) {
    if (!this.hasDifferencesWith(sort)) return;

    this.clear();

    if (!sort.length) return;

    sort.forEach(({ entity, name, order, property }) => {
      const found = this.findByCategory({
        key: entity,
        name,
        property,
      });

      if (found) {
        found.sort = order;

        this.selectedCategories.push(found);
      }
    });

    this.commit();
  }

  private findByCategory(sort: SortIdentifier) {
    return this.categoriesSorts.find((s) => s.id === this.toId(sort));
  }

  private findIndexSelectedCategory(sort: SortIdentifier) {
    return this.selectedCategories.findIndex((s) => s.id === this.toId(sort));
  }

  private toId(sort: SortIdentifier) {
    if (sort.property) return `${sort.key}.${sort.name}.${sort.property}`;

    return `${sort.key}.${sort.name}`;
  }
}
