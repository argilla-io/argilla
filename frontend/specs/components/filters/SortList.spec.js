import { mount } from "@vue/test-utils";
import SortList from "@/components/commons/header/filters/SortList";

function mountSortList() {
  return mount(SortList, {
    propsData: {
      sort: [
        {
          disabled: false,
          group: "Predictions",
          id: "predicted_as",
          key: "predicted_as",
          name: "Predicted as",
          options: Object,
          order: "asc",
          placeholder: "Select labels",
          type: "select",
        }
      ],
      sortOptions: [
        {
          disabled: false,
          group: "Predictions",
          id: "predicted_as",
          key: "predicted_as",
          name: "Predicted as",
          options: Object,
          placeholder: "Select labels",
          selected: undefined,
          type: "select",
        }
      ]
    },
  });
}

describe("SortList", () => {
  let spy = jest.spyOn(console, "error");
  afterEach(() => spy.mockReset());

  test("renders properly", () => {
    const wrapper = mountSortList();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
