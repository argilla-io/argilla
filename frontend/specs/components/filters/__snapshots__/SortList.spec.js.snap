// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortList renders properly 1`] = `
<div>
  <sortfilter sort-options="[object Object]" class="sort"></sortfilter>
  <!---->
  <div class="sort__buttons">
    <re-button class="button-tertiary--small button-tertiary--outline">Cancel</re-button>
    <re-button class="button-primary--small">Sort</re-button>
  </div>
</div>
`;
